### PR TITLE
8336240: Test com/sun/crypto/provider/Cipher/DES/PerformanceTest.java fails with java.lang.ArithmeticException

### PR DESCRIPTION
**Pull request by fguallini**
The manual test Cipher/DES/PerformanceTest.java fails with ArithmeticException due to potential division by zero. The issue arises when calculating the elapsed time using end - start, which could result in zero milliseconds if start and end are identical due to the high speed of execution. This leads to a division error in the following code snippet:

```
start = System.currentTimeMillis();
end = System.currentTimeMillis();
int speed = (int)((data.length * count)/(end - start));
```
This issue is easily reproducible on platforms where System.currentTimeMillis() has low granularity, such as many versions of Windows, end and start can be equal when obtaining System.currentTimeMillis() if the test runs very quickly.

The fix is to use System.nanoTime() instead, which is designed to measure elapsed time with very high precision. This value is then converted to microseconds so the tests can properly calculate the throughput (bytes processed per microsecond) for the report. Example output:

```
Algorithm                      DataSize Rounds Bytes/microsec
DES/ECB/NoPadding               1024    100       66
DES/ECB/NoPadding               1024    1000      50
DES/ECB/NoPadding               8192    100       70
DES/ECB/NoPadding               8192    1000      70
Average:                                            64
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336240](https://bugs.openjdk.org/browse/JDK-8336240): Test com/sun/crypto/provider/Cipher/DES/PerformanceTest.java fails with java.lang.ArithmeticException (**Bug** - P5)

### Reviewers
 * [Bradford Wetmore](https://openjdk.org/census#wetmore) (@bradfordwetmore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20135/head:pull/20135` \
`$ git checkout pull/20135`

Update a local copy of the PR: \
`$ git checkout pull/20135` \
`$ git pull https://git.openjdk.org/jdk.git pull/20135/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20135`

View PR using the GUI difftool: \
`$ git pr show -t 20135`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20135.diff">https://git.openjdk.org/jdk/pull/20135.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20135#issuecomment-2223288398)

### CODE REVIEW

File: test/jdk/com/sun/crypto/provider/Cipher/DES/PerformanceTest.java [hunk 1]
```diff
@@ -185,7 +183,9 @@ public void runTest(byte[] data, int count) throws Exception {
         cipher.doFinal(data, 0, data.length);
         end = System.currentTimeMillis();
 
-        int speed = (int)((data.length * count)/(end - start));
+        // To avoid diving by zero if end is equal to start
```
Comment (rhalade) [inline test/jdk/com/sun/crypto/provider/Cipher/DES/PerformanceTest.java:186]: Test is performing 100 and 1000 rounds of `update`. I don't think we are seeing same start and end. Please test further to determine root cause. Could be the case of overflow if `update` is taking long time.

Comment (fguallini) [reply test/jdk/com/sun/crypto/provider/Cipher/DES/PerformanceTest.java:186]: The root cause is that start and end are equal, leading to the exception ArithmeticException: / by zero. These rounds can execute very quickly; on a macOS platform, they can take as little as 1 millisecond.

This issue is easily reproducible on platforms where System.currentTimeMillis() has low precision, such as many versions of Windows, end and start can be equal when obtaining System.currentTimeMillis() if the rounds run very quickly.

Comment (fguallini) [reply test/jdk/com/sun/crypto/provider/Cipher/DES/PerformanceTest.java]: We can use System.nanoTime() instead, which is designed to measure elapsed time with very high precision. This overcomes the issue on platforms with low granularity, and we should no longer see end - start = 0.

**Review by bradfordwetmore — APPROVED** [review]
While this should work for the systems we use, be aware that it's not guaranteed by the Java API.  From the `System` javadoc:

> public static long nanoTime()
>
> This method provides nanosecond precision, but not necessarily nanosecond resolution (that is, how frequently the value changes) - no guarantees are made except that the resolution is at least as good as that of [currentTimeMillis()](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/System.html#currentTimeMillis()).

It's probably fine, or you could also use the !=0?end-start:1 hack as a backup.

Otherwise, LGTM.

**Comment by rhalade** [comment]
LGTM. Have you run test with iterations to ensure stability. This will be a tier1 test so intermittent failures are not acceptable.

**Comment by bradfordwetmore** [comment]
Alternatively, you could just add 1us and thus force it to be non-zero...it shouldn't affect the `speed` much.  Either way is fine.

**Review by bradfordwetmore — APPROVED** [review]

**Comment by bradfordwetmore** [comment]
> > LGTM. Have you run test with iterations to ensure stability. This will be a tier1 test so intermittent failures are not acceptable.
> 
> 👍 This is a manual test. In any case it's stable, it was successfully ran hundreds of times.

I see it was previously a `/manual` test, before that was an `@ignore`.   The reason it was moved to `@ignore` in 2003 was:

> Exclude this test from regression tests since it is meant to measure performance and not regression testing; it can be run separately if needed.

Is it slow enough these days that we couldn't remove the `/manual`?   How long is the test taking?

@valeriepeng, any reason to keep/not keep this?

**Comment by fguallini** [comment]
> Is it slow enough these days that we couldn't remove the `/manual`? How long is the test taking?

@bradfordwetmore It does not take much time, average ~12 seconds

**Comment by bradfordwetmore** [comment]
> @bradfordwetmore It does not take much time, average ~12 seconds

I could go either way.

**Comment by fguallini** [comment]
> I understand that the main purpose is to provide performance numbers and no one will be watching it, but if it's not taking too much time, it's nice to have the additional test for ensuring nothing went wrong when exercising the code.

Even if someone was watching the report, there is no baseline for comparison that would highlight a lower performance in the current code. So, the only criteria for this test to pass is that it does not timeout. In that case, it can be automatic rather than manual.

**Review by bradfordwetmore — APPROVED** [review]

**Review by bradfordwetmore — APPROVED** [review]

**Comment by bradfordwetmore** [comment]
I double checked for any other mentions of `PerformanceTest.java`, didn't see any.

**Pull request closed without merging**